usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1077,25 +1082,30 @@ zil_lwb_commit(zilog_t *zilog, itx_t *itx, lwb_t *lwb)
 
 	zil_lwb_write_init(zilog, lwb);
 
+cont:
 	/*
 	 * If this record won't fit in the current log block, start a new one.
+	 * For WR_NEED_COPY optimize layout for minimal number of chunks, but
+	 * try to keep wasted space withing reasonable range (12%).
 	 */
-	if (lwb->lwb_nused + reclen + dlen > lwb->lwb_sz) {
+	lwb_sp = lwb->lwb_sz - lwb->lwb_nused;
+	if (reclen > lwb_sp || (reclen + dlen > lwb_sp &&
```

Comment:
"reclen + dlen > lwb_sp"  checks whether record can't fit into lwb completely with all data;
"lwb_sp < ZIL_MAX_LOG_DATA / 8"  checks whether remaining part of lwb significant to bother about record splitting;
"dlen % ZIL_MAX_LOG_DATA == 0 || lwb_sp < reclen + dlen % ZIL_MAX_LOG_DATA"  check whether record split between several lwb's give us residue that could fit nicely into remaining space, not creating extra record/data splits (extra splits require extra zl_get_data() calls).